### PR TITLE
Use service registry builder instead of subclassing

### PR DESCRIPTION
**Pull request by alllex**
This is a refactoring PR that updates service providers to avoid extending `DefaultServiceRegistry`, and instead use the `ServiceRegistryBuilder` to create the registries.

This step is part of a larger goal of making `DefaultServiceRegistry` an implementation detail of Gradle's internal dependency injection framework.

This PR does not cover two remaining places where `DefaultServiceRestiry` is extended:
- `ConnectorServiceRegistry` -- a quick attempt to refactor this resulted in unexpected `NoClassDefFound` errors in cross-version TAPI tests
- `NativeServices` -- it's a singleton that is used as a service registry itself, which makes it non-trivial to refactor into using a builder

### CODE REVIEW

**Comment by alllex** [comment]
@bot-gradle test qfl

**Comment by alllex** [comment]
@bot-gradle test without pts

**Comment by alllex** [comment]
@bot-gradle test this

**Comment by alllex** [comment]
@bot-gradle test this

**Comment by alllex** [comment]
@bot-gradle test this

File: platforms/core-runtime/base-services/src/main/java/org/gradle/internal/service/CloseableServiceRegistry.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.service;
+
+import java.io.Closeable;
+
+public interface CloseableServiceRegistry extends ServiceRegistry, Closeable {
```
Comment (alllex) [inline platforms/core-runtime/base-services/src/main/java/org/gradle/internal/service/CloseableServiceRegistry.java:21]: 💭  In many cases, the code that created a service registry didn't care that it was `DefaultServiceRegistry` specifically. However, at the boundaries between the scopes (like Build Session, Build Tree, etc) the state management needed the ability to close the children registries when the parent is being closed.

**Review by adammurdoch — APPROVED** [review]
Fantastic!

**Comment by alllex** [comment]
@bot-gradle test this

**Pull request merged**